### PR TITLE
Add iOS support to Filament

### PR DESCRIPTION
**Pull request by bejado**
iOS support for Filament on ARM64 devices.

Key changes:
- New platform: `PlatformVkCocoaTouch`
- Build script supports iOS platform and downloads [Apple's ARM toolchain](https://opensource.apple.com/source/clang/clang-800.0.38/src/cmake/platforms/iOS.cmake)
- Added libMoltenVK.dylib ARM64 binary, copied into App bundle at build time

Two commits:
1. Add iOS support to Filament and `build.sh` (use `./build.sh -p ios debug` on Mac host to build)
2. Add a hello-triangle sample iOS app

I haven't had a chance to build on Windows or Linux yet, so I'll wait for Kokoro to succeed before merging.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -346,6 +348,71 @@ function build_android {
...
+    cat build/toolchain-arm64-mac-ios.filament.cmake >> ${TOOLCHAIN_PATH}
+    echo "Successfully downloaded iOS toolchain file and appended Filament-specific settings."
+}
+
+function build_ios_target {
+    local LC_TARGET=`echo $1 | tr '[:upper:]' '[:lower:]'`
+    local ARCH=$2
+
+    echo "Building iOS $LC_TARGET ($ARCH)..."
+    mkdir -p out/cmake-ios-${LC_TARGET}-${ARCH}
```
Comment (prideout) [inline build.sh]: Since there is one and only one arch with iOS, I suppose we could remove the arch suffix here.  On the other hand, this is more consistent with the Android build.

With WASM, we simply create `cmake-webgl-release` and `cmake-webgl-debug`.

File: filament/CMakeLists.txt [hunk 2]
```diff
@@ -325,6 +330,11 @@ target_link_libraries(${TARGET} PUBLIC filaflat)
 target_link_libraries(${TARGET} PUBLIC filabridge)
 target_link_libraries(${TARGET} PUBLIC image_headers)
 
+# iOS does not support OpenGL, and Android links directly against EGL
```
Comment (prideout) [inline filament/CMakeLists.txt]: Is it correct to say "iOS does not support OpenGL" , or is it a specific version of GL?

File: ios/samples/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+libutils.a
+```
+
+The sample application expects these libraries to be present here. Likewise, when building the sample in Release mode, the linker looks for the libraries in `out/ios-release/filament/lib/arm64`. To build Filament in Release mode, replace `debug` with `release` in the above `build.sh` command.
+
+The pre-compiled `libMoltenVK.dylib` library must be present in `third_party/moltenvk/ios`. When building, the Xcode project automatically signs and includes this library inside the application bundle.
+
+## Xcode
+
+Open up the Xcode project at `ios/samples/hello-triangle/hello-triangle.xcodeproj`. Before building you will need to sign in to your Apple developer account and select an appropriate development team in the project editor.
```
Comment (prideout) [inline ios/samples/README.md]: Is there a way to build it using `xcodebuild`?

**Comment by romainguy** [comment]
Mathias made the necessary changes to support es 3.0 only. We should be able to build filament for iOS using GL only.

File: build.sh (hunk 1 above)
Comment (prideout) [reply build.sh]: Looking ahead, I wonder if it would make more sense for the output suffix to be the backend API (vk/mtl/gl) rather than `arm64`.

**Comment by bejado** [comment]
@romainguy- There's still some work to be done to support GL on iOS, mainly due to how Apple implements `CAEAGLLayer` (Apple equivalent of `EGLSurface`). Going to continue looking into it.

File: filament/CMakeLists.txt (hunk 2 above)
Comment (bejado) [reply filament/CMakeLists.txt]: I should rephrase this as "Filament on iOS does not yet support the OpenGL backend" (see my note to Romain below)

File: ios/samples/README.md (hunk 3 above)
Comment (bejado) [reply ios/samples/README.md]: Yes, but unfortunately there’s two issues with that.

1. Users first have to open the editor anyway to select a development team for code signing.
2. `xcodebuild` can only build the app and cannot launch it on the device :(

File: build.sh (hunk 1 above)
Comment (bejado) [reply build.sh]: Yeah, I was following the Android conventions. I think having the architecture here is nice- if I can get OpenGL working, then _in theory_ we could also support the iOS Simulator, in which case the build directory would be `out/cmake-ios-debug-x86-64`. Not opposed to also adding the backend API to the build dir in the future.

File: ios/samples/README.md (hunk 3 above)
Comment (prideout) [reply ios/samples/README.md]: Ah, thanks, understood.

File: build.sh (hunk 1 above)
Comment (prideout) [reply build.sh]: Ah, makes sense!

**Comment by bejado** [comment]
A bit more info: iOS has no concept of a default framebuffer (FBO 0), all OpenGL ES framebuffers must be created with a call to `glGenFramebuffers`. Filament's OpenGL driver though assumes that the default render target is FBO 0 (in `driver.createDefaultRenderTarget()`) so this may need to be reworked.

**Comment by romainguy** [comment]
Yes I remember you mentioning this @bejado. It sounds like we'll need a new `Platform` implementation for iOS and we may have to change those APIs to fix the `FBO 0` assumption. @pixelflinger might have some suggestions there.

**Comment by bejado** [comment]
My current approach is to create a framebuffer and renderbuffer inside of `createDefaultRenderTarget` when on iOS:

```
void OpenGLDriver::createDefaultRenderTarget(
        Driver::RenderTargetHandle rth, int) {
    DEBUG_MARKER()

    construct<GLRenderTarget>(rth);

#if defined(IOS)
    /* Special case on iOS. iOS has no concept of a "default framebuffer", we are forced
     * to create one here. */

    GLRenderTarget* rt = handle_cast<GLRenderTarget*>(rth);

    // Create a framebuffer object
    GLuint framebuffer;
    glGenFramebuffers(1, &framebuffer);

    // Create a renderbuffer object
    GLuint renderbuffer;
    glGenRenderbuffers(1, &renderbuffer);

    rt->gl.color.id = renderbuffer;
    rt->gl.fbo = framebuffer;
#endif

    // nothing more to do
}
```
Then, inside of `PlatformCocoaTouchGL::makeCurrent`, I use the iOS-specific `renderbufferStorage` call to associate the renderbuffer to the `CAEAGLLayer`, which is our `SwapChain`:
```
void PlatformCocoaTouchGL::makeCurrent(SwapChain* drawSwapChain, SwapChain* readSwapChain) noexcept {
    [EAGLContext setCurrentContext:pImpl->mGLContext];

    OpenGLDriver::create(this, sharedGLContext);

    // TODO: assuming renderbuffer and framebuffer are both 1
    CAEAGLLayer* glLayer = (CAEAGLLayer*) drawSwapChain;
    glBindFramebuffer(GL_FRAMEBUFFER, 1);
    glBindRenderbuffer(GL_RENDERBUFFER, 1);
    [pImpl->mGLContext renderbufferStorage:GL_RENDERBUFFER fromDrawable:glLayer];
    glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, 1);
}
```
The problem with this is that I have no way of knowing which framebuffer and renderbuffer to bind (I've hard-coded `1` here, which works, but I don't know if we can always assume this).

One possible solution would be to pass the default renderbuffer to `FSwapChain::makeCurrent`.

```
Renderer.cpp:
    swapChain->makeCurrent(driver, mRenderTarget);
```

Then `OpenGLDriver` could bind the renderbuffer and framebuffer of `mRenderTarget` before calling `mPlatform.makeCurrent`. Then `makeCurrent` just calls the iOS `renderbufferStorage` method.

**Comment by romainguy** [comment]
Can't the `Platform` implementation create the FBO instead? That would solve the problem. The `OpenGLDriver` would only have to query the current framebuffer to know what is the "screen" FBO.

**Comment by romainguy** [comment]
Can you look into adding an iOS Kokoro build before we merge this?
